compiler/next/include/chpl/uast/IndexableLoop.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_INDEXABLELOOP_H
+#define CHPL_UAST_INDEXABLELOOP_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/Loop.h"
+#include "chpl/uast/VariableDecl.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This abstract class represents an indexable loop.
+ */
+class IndexableLoop : public Loop {
+ protected:
+  IndexableLoop(ASTTag tag, ASTList children, int8_t iterandChildNum,
+                int8_t indexVariableDeclStartChildNum,
+                int numIndexVariables,
+                bool usesDo)
+    : Loop(tag, std::move(children),
+           // Compute the start of the loop body.
+           ((iterandChildNum >= 0) + numIndexVariables),
+           usesDo),
+      iterandChildNum_(iterandChildNum),
+      indexVariableDeclStartChildNum_(indexVariableDeclStartChildNum),
+      numIndexVariables_(numIndexVariables) {
+
+    assert(isExpressionASTList(children_));
+  }
+
+  bool indexableLoopContentsMatchInner(const IndexableLoop* other) const {
+    return this->loopContentsMatchInner((const Loop*)other);
+  }
+
+  void indexableLoopMarkUniqueStringsInner(Context* context) const {
+    this->loopMarkUniqueStringsInner(context);
+  }
+
+  int8_t iterandChildNum_;
+  int8_t indexVariableDeclStartChildNum_;
+  int numIndexVariables_;
```

Comment:
This is good for the initial prototype.

In the long run, we want to represent complex indexing expressions, ex. `for (a, (b,c), _) in ....`

So it may be easier to have a single Expression to represent index variable(s). This expression will be either Identifier or Tuple.

Reply:
You're right that this representation does not work for something like `(a, (b,c), _)`, because all nesting is lost. @mppf wanted me to introduce variable declarations for each loop index. Personally I think it would just be easier to leave the expression as the index variable and just interpret the expression appropriately when resolving instead of introducing decls.

Reply:
Would it suffice to just store the original index expression in addition to the index variable declarations? It could be accessed via `indexExpression` and it could be used to type the index variables during resolution (In the `(a, (b, c), _)` case). Perhaps that would be a good compromise.

Reply:
We have tuple variable declarations too; you can write `var (a,b) = (1,2);`. So, I think we need to have some sort of TupleDeclaration uAST node that doesn't exist yet and we should use that here as well. So, I stand by my request that the ForLoop include some sort of Decls for the index variables; the case `for (a, (b,c), _) in ...` will just use `Decls` that we haven't defined yet.

I think it's important that in the common cases along the lines of `for (a,b) in zip(A, B)` that there are no tuples introduced into the representation. Of course there are other cases where we need to handle tuples - but in those cases the compiler is more or less forced into generating code that does things like tuple packing or unpacking. It's much easier to add the tuple representation in later during compilation than to remove it.